### PR TITLE
Change measure function for Hexahedron and the relative test cases.

### PR DESCRIPTION
**Pull request by deltaeecs**
Fix https://github.com/JuliaGeometry/Meshes.jl/issues/193

### CODE REVIEW

File: src/polytopes/hexahedron.jl [hunk 1]
```diff
@@ -28,15 +28,14 @@ function (h::Hexahedron)(u, v, w)
...
-  t4 = Tetrahedron(A1, A2, A3, A7)
-  t5 = Tetrahedron(A2, A3, A4, A7)
-  t6 = Tetrahedron(A2, A3, A4, A8)
-  sum(measure, [t1,t2,t3,t4,t5,t6])
+  t1 = Tetrahedron(A1, A5, A6, A7)
+  t2 = Tetrahedron(A1, A4, A3, A7)
+  t3 = Tetrahedron(A1, A4, A6, A7)
+  t4 = Tetrahedron(A1, A2, A4, A6)
+  t5 = Tetrahedron(A4, A6, A8, A7)
+  sum(measure.([t1,t2,t3,t4,t5]))
```
Comment (juliohm) [inline src/polytopes/hexahedron.jl]: Please preserve the `measure` as the first argument to `sum`.

File: src/polytopes/hexahedron.jl [hunk 2]
```diff
@@ -28,15 +28,14 @@ function (h::Hexahedron)(u, v, w)
 end
 
 function measure(h::Hexahedron)
-  A1, A2, A4, A3,
+  A1, A2, A4, A3,
```
Comment (juliohm) [inline src/polytopes/hexahedron.jl]: Please eliminate this whitespace change.

Comment (deltaeecs) [reply src/polytopes/hexahedron.jl]: haha,ok👌

Comment (juliohm) [inline src/polytopes/hexahedron.jl]:
```suggestion
  A1, A2, A4, A3,
```

File: src/polytopes/hexahedron.jl (hunk 1 above)
Comment (deltaeecs) [reply src/polytopes/hexahedron.jl]: ok

File: src/polytopes/hexahedron.jl [hunk 3]
```diff
@@ -30,13 +30,12 @@ end
...
-  t4 = Tetrahedron(A1, A2, A3, A7)
-  t5 = Tetrahedron(A2, A3, A4, A7)
-  t6 = Tetrahedron(A2, A3, A4, A8)
-  sum(measure, [t1,t2,t3,t4,t5,t6])
+  t1 = Tetrahedron(A1, A5, A6, A7)
+  t2 = Tetrahedron(A1, A4, A3, A7)
+  t3 = Tetrahedron(A1, A4, A6, A7)
+  t4 = Tetrahedron(A1, A2, A4, A6)
+  t5 = Tetrahedron(A4, A6, A8, A7)
+  sum(measure.([t1,t2,t3,t4,t5]))
```
Comment (juliohm) [inline src/polytopes/hexahedron.jl]:
```suggestion
  sum(measure, [t1,t2,t3,t4,t5])
```

**Comment by juliohm** [comment]
I've implemented the proposed fixes directly on GitHub. I will wait for the tests and tomorrow morning I can double check it again before we merge it 👍🏽 Thanks for the contribution!

**Comment by deltaeecs** [comment]
> I've implemented the proposed fixes directly on GitHub. I will wait for the tests and tomorrow morning I can double check it again before we merge it 👍🏽 Thanks for the contribution!

Thx, have a good night!

File: test/polytopes.jl [hunk 4]
```diff
@@ -139,13 +139,12 @@
     h = Hexahedron(P3[(0,0,0),(1,0,0),(1,1,0),(0,1,0),
                       (0,0,1),(1,0,1),(1,1,1),(0,1,1)])
     @test volume(h) ≈ T(1*1*1)
-    h = Hexahedron(P3[(0,0,0),(2,0,0),(2,2,0),(0,2,0),
+    h = Hexahedron(P3[(0,0,0),(1,0,0),(1,1,0),(0,1,0),
                       (0,0,2),(2,0,2),(2,2,2),(0,2,2)])
-    @test volume(h) ≈ T(2*2*2)
```
Comment (juliohm) [inline test/polytopes.jl]: Please keep the previous test. We only add tests or fix tests that are not correct.

File: test/polytopes.jl [hunk 5]
```diff
@@ -139,13 +139,12 @@
...
+    @test volume(h) ≈ T((1 + 4 + sqrt(1*4))*2/3)
```
Comment (juliohm) [inline test/polytopes.jl]: Need to explain where this formula is coming from so that future readers can quickly understand the test.

**Review by juliohm — CHANGES_REQUESTED** [review]
Attached a first round of reviews 👍🏽

File: test/polytopes.jl [hunk 6]
```diff
@@ -142,10 +142,12 @@
     h = Hexahedron(P3[(0,0,0),(2,0,0),(2,2,0),(0,2,0),
                       (0,0,2),(2,0,2),(2,2,2),(0,2,2)])
     @test volume(h) ≈ T(2*2*2)
-    h = Hexahedron(P3[(0,0,0),(1,0,0),(1,1,0),(0,1,0),
-                      (0,0,1),(1,0,1),(1,1,2),(0,1,1)])
-    t = Tetrahedron(P3[(1,0,1),(1,1,1),(0,1,1),(1,1,2)])
-    @test volume(h) ≈ T(1*1*1) + volume(t)
+    #=  Volume formula of a frustum of a prism is V = 1/3*H*(S₁+S₂+sqrt(S₁*S₂))
+        Here we build a hexahedron which is a frustum of a prism with 
+        buttom area S₁=4, top area S₂=1, hight H = 2      =#
```
Comment (juliohm) [inline test/polytopes.jl]:
```suggestion
    # volume formula of a frustum of a prism is V = 1/3*H*(S₁+S₂+sqrt(S₁*S₂))
    # here we build a hexahedron which is a frustum of a prism with
    # bottom area S₁= 4, top area S₂= 1, height H = 2
```

File: test/polytopes.jl [hunk 7]
```diff
@@ -142,10 +142,12 @@
     h = Hexahedron(P3[(0,0,0),(2,0,0),(2,2,0),(0,2,0),
                       (0,0,2),(2,0,2),(2,2,2),(0,2,2)])
     @test volume(h) ≈ T(2*2*2)
-    h = Hexahedron(P3[(0,0,0),(1,0,0),(1,1,0),(0,1,0),
-                      (0,0,1),(1,0,1),(1,1,2),(0,1,1)])
-    t = Tetrahedron(P3[(1,0,1),(1,1,1),(0,1,1),(1,1,2)])
-    @test volume(h) ≈ T(1*1*1) + volume(t)
+    # volume formula of a frustum of a prism is V = 1/3*H*(S₁+S₂+sqrt(S₁*S₂))
```
Comment (juliohm) [inline test/polytopes.jl:146]:
```suggestion

    # volume formula of a frustum of a prism is V = 1/3*H*(S₁+S₂+sqrt(S₁*S₂))
```

**Comment by juliohm** [comment]
For future reference, it is easy to visualize these objects and share these plots to help the reviewer:

```julia
using Meshes, MeshViz
import GLMakie

h = Hexahedron(Point3[(0,0,0),(2,0,0),(2,2,0),(0,2,0), (0,0,2),(1,0,2),(1,1,2),(0,1,2)])

viz(h)
```
image

**Comment by juliohm** [comment]
Can you please confirm that the previous code without changes was producing incorrect results with this frustum?

**Comment by juliohm** [comment]
I am trying to separate the bug in the test from the bug in the code. I understand the test was incorrect, but perhaps the original code was already ok? Appreciate if you can double check that.

**Comment by deltaeecs** [comment]
> Can you please confirm that the previous code without changes was producing incorrect results with this frustum?

I just tested again with
```
using Meshes
P3 = Point3
h = Hexahedron(P3[  (0,0,0),(2,0,0),(2,2,0),(0,2,0),
                    (0,0,2),(1,0,2),(1,1,2),(0,1,2)])

function measureOld(h::Hexahedron)
    A1, A2, A4, A3,
    A5, A6, A8, A7 = h.vertices
    t1 = Tetrahedron(A1, A2, A5, A7)
    t2 = Tetrahedron(A2, A6, A5, A7)
    t3 = Tetrahedron(A2, A6, A7, A8)
    t4 = Tetrahedron(A1, A2, A3, A7)
    t5 = Tetrahedron(A2, A3, A4, A7)
    t6 = Tetrahedron(A2, A3, A4, A8)
    sum(measure, [t1,t2,t3,t4,t5,t6])
end

1/3*2*(1 + 4 + sqrt(1*4)) # output 4.666666666666666
measure(h)  # output 4.666666666666666
measureOld(h) # output 5.333333333333333
```

**Comment by deltaeecs** [comment]
> I am trying to separate the bug in the test from the bug in the code. I understand the test was incorrect, but perhaps the original code was already ok? Appreciate if you can double check that.

Using the origin code, tetrahedrons we get will overlap in the hexahedron when hexahedron is stretched. Since points 1,2,7,8 could not be in the same plane.

**Comment by juliohm** [comment]
Awesome, thanks for checking 💯

**Comment by deltaeecs** [comment]
> I am trying to separate the bug in the test from the bug in the code. I understand the test was incorrect, but perhaps the original code was already ok? Appreciate if you can double check that.

I'd like to visualize the overlap but my GLMakie never worked.😥 You can visualize it by making two meshes of [t1, t2, t3] and [t4, t5, t6] and visualize them separately.

**Comment by juliohm** [comment]
Thanks for another great fix! 💯

**Pull request merged**